CAToolsetPerfDataCollector/PerfDataCollector.cpp

```diff
@@ -0,0 +1,306 @@
+    // BE Pass
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetPassDuration<BackEndPass, &PerfData::BEPass>);
+
+    //
+    // Activities in FE during CA Pass
+    //
+
+    // AST Creation
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<AstCreation, &PerfData::ASTCreation>);
+    // SST Clients, e.g., PREfast
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<CodeAnalysisPlugins, &PerfData::ASTClients>);
+
+    //
+    // Activities in PREfast during CA Pass
+    //
+
+    // CA PREfast's Function Analysis (including its own plug-ins, EspXEngine, and EspXEngine extensions)
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<CodeAnalysisFunction, &PerfData::CAFunction>);
+    // PREfast's FPA Function Analysis (FPA = Function Path Analysis)
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<CodeAnalysisPREfastFpaFunction, &PerfData::FPAFunction>);
+    // EspXEngine's CFG Building
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<EspXEngineCfgBuild, &PerfData::EspXCfgBuild>);
+    // EspXEngine's All Checks (BufferCheck and extensions, whichever are enabled)
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<EspXEngineChecks, &PerfData::EspXAllChecks>);
+    // EspXEngine's PathSensitiveChecks (BufferCheck and some of the extensions, whichever are enabled)
+    MatchEventStackInMemberFunction(eventStack, this,
+        &PerfDataCollector::GetCodeAnalysisEventDuration<EspXEnginePathSimulation, &PerfData::EspXPathSensitiveChecks>);
+
+    // Tell the analysis driver to proceed to the next event
+    return AnalysisControl::CONTINUE;
+}
+
+// Prints the per-TU performance data in CSV format. The first column is the file path.
+// The rest of the columns are the performance data in microseconds.
+void PerfDataCollector::PrintRecords()
+{
+    constexpr auto allZero = PerfData{};
+
+    // Print Header
+    std::wcout << L"File Path, " << PerfData::PerfDataHeader << std::endl;
+
+    for (const auto& [pathId, data] : perfDataPerTu)
+    {
+        // If CA Pass is less than FE Pass, it is for files that are not target of Code Analysis,
+        // or CA Pass is not captured properly - e.g., due to errors only occurring during CA Pass.
+        // or exceptions during Code Analysis.
+        // Ignore the record by reporting everything as 0. We still want the TU to be listed.
+        const auto& d = (data.CAPass < data.FEPass) ? allZero : data;
+
+        std::wcout
+            << *pathId.path << L", " << d.FEPass.count() << L", " << d.BEPass.count() << L", " << d.CAPass.count() << L", "
```

Comment:
Minor: since this project is using C++20, we could use `std::format`.

Reply:
Leaving it as is if that is OK. These execute just once per invocation.